### PR TITLE
CASSANDRA-17154 trunk: Guardrail to disable read-before-write list operations

### PR DESCRIPTION
**Pull request by adelapena**

### CODE REVIEW

File: conf/cassandra.yaml [hunk 1]
```diff
@@ -1607,3 +1607,6 @@ enable_drop_compact_storage: false
 #         disallowed: []
 # Guardrail to allow/disallow user-provided timestamps. Defaults to true.
 #     user_timestamps_enabled: true
+# Guardrail to allow/disallow list operations that require read before write, i.e. setting list element by index and
+# removing list elements by either index or value. Defaults to true.
+#     read_before_write_list_operations_enabled: true
```
Comment (ekaterinadimitrova2) [inline conf/cassandra.yaml:1617]: Shouldn't it be disabled by default? Same as every new feature we add?

Comment (ekaterinadimitrova2) [reply conf/cassandra.yaml:1617]: My bad, they are disabled globally

File: test/unit/org/apache/cassandra/db/guardrails/GuardrailReadBeforeWriteListOperationsTest.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+                          () -> "UPDATE %s SET l[1] = 4 WHERE k = 0",
+                          () -> "DELETE l[1] FROM %s WHERE k = 0",
+                          () -> "INSERT INTO %s (k, l) VALUES (0, [1, 2, 3])",
+                          () -> "UPDATE %s SET l = l - [2] WHERE k = 0",
+                          () -> "BEGIN BATCH UPDATE %s SET l[1] = 0 WHERE k = 0; APPLY BATCH",
+                          () -> "BEGIN BATCH DELETE l[1] FROM %s WHERE k = 0; APPLY BATCH",
+                          () -> "BEGIN BATCH UPDATE %s SET l = l - [3] WHERE k = 0; APPLY BATCH");
+    }
+
+    private void assertRows(Object[]... rows) throws Throwable
```
Comment (ekaterinadimitrova2) [inline test/unit/org/apache/cassandra/db/guardrails/GuardrailReadBeforeWriteListOperationsTest.java]: nit: I think this method should be after the one that uses it, `testGuardrail(String query, String expectedMessage, Object[]... rows) throws Throwable`

File: conf/cassandra.yaml (hunk 1 above)
Comment (adelapena) [reply conf/cassandra.yaml:1617]: The guardrail is also disabled locally. Here `true` means that read before write operations are enabled as usual, so the guardrail isn’t having any effect.

File: src/java/org/apache/cassandra/cql3/Lists.java [hunk 3]
```diff
@@ -565,6 +568,8 @@ public void execute(DecoratedKey partitionKey, UpdateParameters params) throws I
         {
             assert column.type.isMultiCell() : "Attempted to delete from a frozen list";
 
+            Guardrails.readBeforeWriteListOperationsEnabled.ensureEnabled("Removal of list items requiring read before write", params.clientState);
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/cql3/Lists.java]:
```suggestion
            Guardrails.readBeforeWriteListOperationsEnabled
            .ensureEnabled("Removal of list items requiring read before write", params.clientState);
```

File: src/java/org/apache/cassandra/cql3/Lists.java [hunk 4]
```diff
@@ -602,6 +607,9 @@ public boolean requiresRead()
         public void execute(DecoratedKey partitionKey, UpdateParameters params) throws InvalidRequestException
         {
             assert column.type.isMultiCell() : "Attempted to delete an item by index from a frozen list";
+
+            Guardrails.readBeforeWriteListOperationsEnabled.ensureEnabled("Removal of list items by index requiring read before write", params.clientState);
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/cql3/Lists.java]:
```suggestion
            Guardrails.readBeforeWriteListOperationsEnabled
            .ensureEnabled("Removal of list items by index requiring read before write", params.clientState);
```

File: src/java/org/apache/cassandra/cql3/Lists.java [hunk 5]
```diff
@@ -448,6 +449,8 @@ public void execute(DecoratedKey partitionKey, UpdateParameters params) throws I
             // we should not get here for frozen lists
             assert column.type.isMultiCell() : "Attempted to set an individual element on a frozen list";
 
+            Guardrails.readBeforeWriteListOperationsEnabled.ensureEnabled("Setting of list items by index requiring read before write", params.clientState);
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/cql3/Lists.java]:
```suggestion
            Guardrails.readBeforeWriteListOperationsEnabled
            .ensureEnabled("Setting of list items by index requiring read before write", params.clientState);
```

**Review by ekaterinadimitrova2 — COMMENTED** [review]
Looks good to me, I left only a few tiny formatting suggestions. Rebase and last CI run and I think it is ready to commit.

**Pull request closed without merging**